corehq/apps/sms/models.py

```diff
@@ -656,3 +690,487 @@ def opt_out_sms(cls, phone_number):
+        ERROR_NO_TWO_WAY_PHONE_NUMBER:
+            ugettext_noop('Contact has no two-way phone number.'),
+        ERROR_INVALID_CUSTOM_CONTENT_HANDLER:
+            ugettext_noop('Invalid custom content handler.'),
+        ERROR_CANNOT_LOAD_CUSTOM_CONTENT_HANDLER:
+            ugettext_noop('Cannot load custom content handler.'),
+        ERROR_CANNOT_FIND_FORM:
+            ugettext_noop('Cannot find form.'),
+        ERROR_FORM_HAS_NO_QUESTIONS:
+            ugettext_noop('No questions were available in the form. Please '
+                'check that the form has questions and that display conditions '
+                'are not preventing questions from being asked.'),
+        ERROR_CASE_EXTERNAL_ID_NOT_FOUND:
+            ugettext_noop('The case with the given external ID was not found.'),
+        ERROR_MULTIPLE_CASES_WITH_EXTERNAL_ID_FOUND:
+            ugettext_noop('Multiple cases were found with the given external ID.'),
+        ERROR_NO_CASE_GIVEN:
+            ugettext_noop('The form requires a case but no case was provided.'),
+        ERROR_NO_EXTERNAL_ID_GIVEN:
+            ugettext_noop('No external ID given; please include case external ID after keyword.'),
+        ERROR_COULD_NOT_PROCESS_STRUCTURED_SMS:
+            ugettext_noop('Error processing structured SMS.'),
+        ERROR_SUBEVENT_ERROR:
+            ugettext_noop('View details for more information.'),
+        ERROR_TOUCHFORMS_ERROR:
+            ugettext_noop('An error occurred in the formplayer service.'),
+        ERROR_INTERNAL_SERVER_ERROR:
+            ugettext_noop('Internal Server Error'),
+        ERROR_GATEWAY_ERROR:
+            ugettext_noop('Gateway error.'),
+        ERROR_NO_SUITABLE_GATEWAY:
+            ugettext_noop('No suitable gateway could be found.'),
+    }
+
+    domain = models.CharField(max_length=126, null=False, db_index=True)
+    date = models.DateTimeField(null=False, db_index=True)
+    source = models.CharField(max_length=3, choices=SOURCE_CHOICES, null=False)
+    source_id = models.CharField(max_length=126, null=True)
+    content_type = models.CharField(max_length=3, choices=CONTENT_CHOICES, null=False)
+
+    # Only used when content_type is CONTENT_SMS_SURVEY or CONTENT_IVR_SURVEY
+    # This is redundantly stored here (as well as the subevent) so that it
+    # doesn't have to be looked up for reporting.
+    form_unique_id = models.CharField(max_length=126, null=True)
+    form_name = models.TextField(null=True)
+
+    # If any of the MessagingSubEvent status's are STATUS_ERROR, this is STATUS_ERROR
+    status = models.CharField(max_length=3, choices=STATUS_CHOICES, null=False)
+    error_code = models.CharField(max_length=126, null=True)
+    additional_error_text = models.TextField(null=True)
+    recipient_type = models.CharField(max_length=3, choices=RECIPIENT_CHOICES, null=True, db_index=True)
+    recipient_id = models.CharField(max_length=126, null=True, db_index=True)
```

Comment:
I only see a single query that uses this class at the moment (perhaps there are more to come):
```
data = MessagingEvent.objects.filter(
            Q(domain=self.domain),
            Q(date__gte=self.datespan.startdate_utc),
            Q(date__lte=self.datespan.enddate_utc),
            (Q(source__in=source_filter) |
                Q(content_type__in=content_type_filter) |
                Q(messagingsubevent__content_type__in=content_type_filter)),
        ).distinct()
```

Given that query it would make sense to have a composite index on `domain, date` rather than individual ones. Also the indexes on `recipient_type` and `recipient_id` seem unused.